src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBiomartRetriever.java

```diff
@@ -0,0 +1,232 @@
+
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+import org.reactome.addlinks.dataretrieval.FileRetriever;
+
+import java.io.*;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.net.URLEncoder;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.*;
+
+public class EnsemblBiomartRetriever extends FileRetriever {
+
+    public EnsemblBiomartRetriever() { }
+
+    public EnsemblBiomartRetriever(String retrieverName)
+    {
+        super(retrieverName);
+    }
+
+    private static final int UNIPROT_QUERY_TOTAL = 2;
+    private static int successCount;
+    private static final String  microarrayTypesBaseQuery = "type=listAttributes&mart=ENSEMBL_MART_ENSEMBL&virtualSchema=default&dataset=BIOMART_SPECIES_NAME_gene_ensembl&interface=default&attributePage=feature_page&attributeGroup=external&attributeCollection=microarray";
+
+    /**
+     * Downloads Ensembl-Microarray and Ensembl-Uniprot identifier mapping files for all species, if they exist.
+     * @throws Exception
+     */
+    protected void downloadData() throws Exception {
+
+        // TODO: Add unit tests
+        // Get names of all organisms we add links and/or microarray data for.
+        // Species names are in biomart format (eg: hsapiens).
+        for (String biomartSpeciesName : getBiomartSpeciesNames()) {
+            logger.info("Retrieving Biomart files for " + biomartSpeciesName);
+            // Create directory where Biomart files will be stored.
+            new File(this.destination).mkdirs();
+
+            logger.info("Retrieving microarray data");
+            // Query Biomart for existing microarray 'types' (not ids) that exist for this species.
+            Set<String> microarrayTypes = queryBiomart(getMicroarrayTypesQuery(biomartSpeciesName), 0);
+            // Iterate through each microarray type and retrieve Ensembl-Microarray identifier mappings.
+            // All mappings are stored in a single file, (eg: hsapiens_microarray);
+            retrieveAndStoreMicroarrayData(biomartSpeciesName, getBiomartXMLPath(), microarrayTypes);
+
+            // Query Ensembl-Uniprot (swissprot and trembl) identifier mapping data from Biomart
+            // and write it to a file (eg: hsapiens_uniprot).
+            logger.info("Retrieving UniProt data");
+            retrieveAndStoreUniprotData(biomartSpeciesName, getBiomartXMLPath());
+
+            logger.info("Completed Biomart data retrieval for " + biomartSpeciesName);
+        }
+    }
+
+    /**
+     * Query Biomart for both Uniprot mapping types (curated -- SwissProt, generated -- TrEMBL).
+     * @param biomartSpeciesName
+     * @throws Exception
+     */
+    private void retrieveAndStoreUniprotData(String biomartSpeciesName, String biomartXMLPath) throws IOException {
```

Comment:
For modularity, it might be good to split this method into two - one to retrieve the UniProt data and one to store it.  It will make unit testing easier if you decide to test this code.

Reply:
Done